comms/src/outbound_message_service/outbound_message_service.rs

```diff
@@ -0,0 +1,318 @@
+        zmq::{Context, InprocAddress, ZmqEndpoint, ZmqError},
+    },
+    outbound_message_service::{broadcast_strategy::BroadcastStrategy, outbound_message::OutboundMessage},
+    peer_manager::node_identity::NodeIdentity,
+    types::{Challenge, MESSAGE_PROTOCOL_VERSION, WIRE_PROTOCOL_VERSION},
+};
+use derive_error::Error;
+use digest::Digest;
+use rand::{CryptoRng, Rng};
+use rmp_serde;
+use serde::Serialize;
+use std::{ops::Mul, sync::Arc};
+use tari_crypto::{
+    keys::{PublicKey, SecretKey},
+    signatures::{SchnorrSignature, SchnorrSignatureError},
+};
+use tari_utilities::{chacha20, ByteArray, ByteArrayError, Hashable};
+
+#[derive(Debug, Error)]
+pub enum OutboundError {
+    /// Problem setting up a socket to an outbound message pool
+    SocketError(ZmqError),
+    /// Could not connect to the outbound message pool
+    SocketConnectionError(zmq::Error),
+    /// Problem sending message to outbound message pool
+    SendError(ConnectionError),
+    /// The secret key was not defined in the node identity
+    UndefinedSecretKey,
+    /// The message signature could not be serialized to a vector of bytes
+    SignatureSerializationError,
+    /// The generated shared secret could not be serialized to a vector of bytes
+    SharedSecretSerializationError(ByteArrayError),
+    /// The message could not be serialized
+    MessageSerializationError(MessageError),
+    /// Could not successfully sign the message
+    SignatureError(SchnorrSignatureError),
+}
+
+/// Handler functions use the OutboundMessageService to send messages to peers. The OutboundMessage service will receive
+/// messages from handlers, apply a broadcasting strategy, encrypted and serialized the messages into OutboundMessages
+/// and write them to the outbound message pool.
+pub struct OutboundMessageService<PubKey, SecKey> {
+    context: Context,
+    outbound_address: InprocAddress,
+    node_identity: Arc<NodeIdentity<PubKey, SecKey>>,
+}
+
+impl<PubKey, SecKey> OutboundMessageService<PubKey, SecKey>
+where
+    PubKey: PublicKey<K = SecKey> + Hashable,
+    SecKey: SecretKey + Mul<PubKey, Output = PubKey> + Mul<Output = SecKey> + Serialize,
+{
+    /// Constructs a new OutboundMessageService from the context, node_identity and outbound_address
+    pub fn new(
+        context: Context,
+        outbound_address: InprocAddress,
+        node_identity: Arc<NodeIdentity<PubKey, SecKey>>,
+    ) -> OutboundMessageService<PubKey, SecKey>
+    {
+        OutboundMessageService {
+            context,
+            outbound_address,
+            node_identity,
+        }
+    }
+
+    /// Encrypt the message_envelope_body with the generated shared secret if the Encrypted IdentityFlag is set
+    fn encrypt_envelope_body(
+        &self,
+        flags: IdentityFlags,
+        message_envelope_body: &Frame,
+        dest_node_public_key: &PubKey,
+    ) -> Result<Frame, OutboundError>
+    {
+        if flags.contains(IdentityFlags::ENCRYPTED) {
```

Comment:
My feeling is that `encrypt-envelope_body` should do what it says on the box; i.e. this check is better made in the calling function?

You also save yourself from passing in `flags`

Reply:
Check moved to calling function making ```encrypt_envelope_body``` only perform the encryption.